### PR TITLE
Fix error decoding pico-8 GIFs (fix #3922)

### LINKED ISSUE #3922: Error decoding pico-8 GIFs
Problems decoding gif files created from pico-8.

Reported here https://twitter.com/calixjumio/status/1667223925292408833
Internal ticket: https://igarastudio.zendesk.com/agent/tickets/4880

I have created a GIF to with pico-8 to reproduce this bug:

untitled_4

### PR DESCRIPTION
**Pull request by Gasparoken**
Tested loading several gif files. Representation of gif files was improved. The pico-8 sample and others cases are fixed.

### CODE REVIEW

**Comment by dacap** [comment]
I've dedicated some time to test this change with several GIF files but I'll need more time. Meanwhile please @martincapello could you please take a look to the commit and check if you can understand the changes (and the comments).

File: src/app/file/gif_format.cpp [hunk 1]
```diff
@@ -493,9 +494,12 @@ class GifDecoder {
...
   // Adds colors used in the GIF frame so we can draw it over
-  // m_currentImage. If the frame contains a local colormap, we try to
-  // find them in the current sprite palette (using
-  // Palette::findExactMatch()) so we don't add duplicated entries.
+  // m_currentImage. If the frame 0 contains global palette and no local palette
+  // we'll take the entire global palette as a pelette, no matter if it's
+  // used or not in the image.
+  // On the other hand if the frame contains local palette, the colors will be
+  // added to the palette of the sprite via Palette::findExactMatch()) so
+  // we don't add duplicated entries.
```
Comment (dacap) [inline src/app/file/gif_format.cpp:502]: You can reword/spellcheck this comment (and I think frames doesn't contain the global colormap, but the file is the one with the global palette, frames contain local colormaps).

File: src/app/file/gif_format.cpp [hunk 2]
```diff
@@ -508,161 +512,136 @@ class GifDecoder {
...
-      }
+    if (m_frameNum == 0 && !m_opaque &&
+        m_localTransparentIndex < ncolors &&
+        m_localTransparentIndex >= 0) {
+      m_bgIndex = m_localTransparentIndex;
+      m_sprite->setTransparentColor(m_bgIndex);
+      clear_image(m_currentImage.get(), m_bgIndex);
+      clear_image(m_previousImage.get(), m_bgIndex);
+      usedEntries[m_bgIndex] = true;
+    }
```
Comment (dacap) [inline src/app/file/gif_format.cpp:523]: Probably I'd add a comment about this (why the m_bgIndex changed to the local transparent index), as the old comment was lost in the change.

File: src/app/file/gif_format.cpp [hunk 3]
```diff
@@ -508,161 +512,136 @@ class GifDecoder {
...
-        !m_opaque &&
-        needCheckLocalTransparent) {
-      for (const auto& i : LockImageBits<IndexedTraits>(frameImage)) {
-        if (i == m_bgIndex) {
-          needsExtraBgColor = true;
-          break;
-        }
+    std::unique_ptr<Palette> palette;
+    if (m_frameNum == 0) {
+      int extraColor = m_localTransparentIndex >= ncolors ? 1 : 0;
```
Comment (dacap) [inline src/app/file/gif_format.cpp:540]:
```suggestion
      int extraColor = (m_localTransparentIndex >= ncolors ? 1 : 0);
```

File: src/app/file/gif_format.cpp [hunk 4]
```diff
@@ -508,161 +512,136 @@ class GifDecoder {
...
+      palette.reset(new Palette(m_frameNum, usedNColors + extraColor));
```
Comment (dacap) [inline src/app/file/gif_format.cpp:541]:
```suggestion
      palette = std::make_unique<Palette>(m_frameNum, usedNColors + extraColor);
```

File: src/app/file/gif_format.cpp [hunk 5]
```diff
@@ -508,161 +512,136 @@ class GifDecoder {
...
+      if (!m_opaque && m_localTransparentIndex >= ncolors) {
+        // If m_localTransparentIndex is out of bounds, we'll define
+        // m_localTransparentIndex = last used color;
```
Comment (dacap) [inline src/app/file/gif_format.cpp:544]: Can the local transparent index be out of bounds (bounds of the local color map I guess?)? Why this happens? (buggy GIFs? maybe we can explain that some GIF files contains this kind of "out of bounds" transparent index).

File: src/app/file/gif_format.cpp [hunk 6]
```diff
@@ -508,161 +512,136 @@ class GifDecoder {
...
+        m_localTransparentIndex = palette->size() - 1;
+        m_bgIndex = m_localTransparentIndex;
+        m_sprite->setTransparentColor(m_bgIndex);
+        clear_image(m_currentImage.get(), m_bgIndex);
+        clear_image(m_previousImage.get(), m_bgIndex);
+        m_relocatedBgIndex = true;
```
Comment (dacap) [inline src/app/file/gif_format.cpp:550]: We are doing something similar above, probably could be a new member function/method this?

File: src/app/file/gif_format.cpp [hunk 7]
```diff
@@ -508,161 +512,136 @@ class GifDecoder {
...
+          j = palette->size() - 1;
+          palette->setEntry(j, localTransIndexColor);
+        }
+        if (j < 256) {
+          m_remap.map(m_bgIndex, j);
+          m_remap.map(m_localTransparentIndex, m_bgIndex);
+        }
       }
-      m_remap.map(i, j);
+      int bgIndex = palette->size() >= 256 ? m_bgIndex : m_remap[m_bgIndex];
```
Comment (dacap) [inline src/app/file/gif_format.cpp:641]:
```suggestion
      int bgIndex = (palette->size() >= 256 ? m_bgIndex : m_remap[m_bgIndex]);
```

File: src/app/file/gif_format.cpp [hunk 8]
```diff
@@ -890,6 +870,11 @@ class GifDecoder {
   DisposalMethod m_disposalMethod;
   int m_bgIndex;
   int m_localTransparentIndex;
+  // When the local transparent index is out of bounds in frame 0, we need relocation of
+  // the global transparent index (m_bgIndex) and it was defined to move to the last index.
+  // Then if m_relocatedBgIndex == true, any pixel == m_bgIndex should be remaped to
+  // a new color, options: find a color match, or add a new color to the palette.
+  bool m_relocatedBgIndex;
```
Comment (dacap) [inline src/app/file/gif_format.cpp:877]: Reword the comment.

File: src/app/file/gif_format.cpp [hunk 9]
```diff
@@ -508,161 +512,136 @@ class GifDecoder {
...
-      // be picked from the sprite palette, instead of m_firstLocalColorMap.
-      if (m_firstLocalColormap && m_firstLocalColormap->ColorCount > j) {
-        // We need add this extra color to m_firstLocalColormap, because
-        // it might has not been considered in the first getFrameColormap execution.
-        // (this happen when: in the first execution of getFrameColormap function
-        // an extra color was not needed)
-        m_firstLocalColormap->Colors[j].Red = rgba_getr(palette->getEntry(j));
-        m_firstLocalColormap->Colors[j].Green = rgba_getg(palette->getEntry(j));
-        m_firstLocalColormap->Colors[j].Blue = rgba_getb(palette->getEntry(j));
+    if (!stopRemap) {
```
Comment (dacap) [inline src/app/file/gif_format.cpp:613]: Probably we can add some comments to this logic.

**Review by dacap — CHANGES_REQUESTED** [review]

File: src/app/file/gif_format.cpp (hunk 5 above)
Comment (Gasparoken) [reply src/app/file/gif_format.cpp:544]: Screenshot 2023-07-03 at 17 33 48

File: src/app/file/gif_format.cpp (hunk 6 above)
Comment (Gasparoken) [reply src/app/file/gif_format.cpp:550]: I'll use a small lambda function.

Comment (dacap) [reply src/app/file/gif_format.cpp:550]: It's better to create a member function if we are inside a class.

File: src/app/file/gif_format.cpp (hunk 8 above)
Comment (Gasparoken) [reply src/app/file/gif_format.cpp:877]: I made a new PR: https://github.com/aseprite/aseprite/pull/3959
(keeping this comment in mind)

**Pull request closed without merging**